File: Duyệt - Vét cạn/BRUOLANG.cpp
#include <iostream>
#include <algorithm>
#define llong long long

using namespace std;

struct Data {
    int kind, x, y;
    llong w;
} b[100006];

llong a[100006], S[100006], c[100006], d[100006];
int n, m, k;

template<typename T>
bool fast_scan(T &num) {
    num = 0;
    register char c = getchar();
    while (c != '-' && (c < '0' || '9' < c)) {
        if (feof(stdin)) return false;
        c = getchar();
    }
    bool neg = false;
    if (c == '-') {
        neg = true;
        c = getchar();
    }
    for(; '0' <= c && c <= '9'; c = getchar()) num = (num << 1) + (num << 3) + (c - '0');
    if (neg) num = -num;
    return true;
}

int main() {
    fast_scan(n), fast_scan(m), fast_scan(k);
    for(int i=1; i<=n; ++i) fast_scan(a[i]);
    for(int i=1; i<=m; ++i) {
        fast_scan(b[i].kind), fast_scan(b[i].x), fast_scan(b[i].y), fast_scan(b[i].w);
        if (b[i].kind == 2) b[i].w *= (-1);
    }
    for(int i=1; i<=k; ++i) {
        int kind, x, y; fast_scan(kind), fast_scan(x), fast_scan(y);
        ++c[x], --c[y+1];
    }
    for(int i=1; i<=m; ++i) {
        c[i] += c[i-1];
        d[b[i].x] += c[i] * b[i].w, d[b[i].y+1] -= c[i] * b[i].w;
    }
    for(int i=1; i<=n; ++i) {
        d[i] += d[i-1];
        a[i] += d[i];
        S[i] = S[i-1] + a[i];
    }
    int u1, v1, u2, v2;
    llong S1, S2;
    fast_scan(u1), fast_scan(v1), fast_scan(S1);
    fast_scan(u2), fast_scan(v2), fast_scan(S2);
    if (abs(S[v1] - S[u1-1] - S1) < abs(S[v2] - S[u2-1] - S2)) cout << "SAN WIN";
    else if (abs(S[v1] - S[u1-1] - S1) > abs(S[v2] - S[u2-1] - S2)) cout << "MI WIN";
    else cout << "HOA";
    return 0;
}
    

File: Duyệt - Vét cạn/BRUOLANG.pas
var         n, m, k, i, j: longint;
	        kind, l, r: longint;
	        u1, u2, v1, v2: longint;
            x, y, c: array[0..100006] of longint;
            a, d, w, S: array[0..100006] of int64;
            res, S1, S2: int64;

begin
            read(n, m, k); c[0] := 0; d[0] := 0; S[0] := 0;
            for i:=1 to n do read(a[i]);
            for i:=1 to m do
                begin
                    read(kind, x[i], y[i], w[i]);
                    if kind = 2 then w[i] := -w[i];
                end;
            for i:=1 to k do
                begin
                    read(kind, l, r);
                    inc(c[l]); dec(c[r+1]);
                end;
            for i:=1 to m do
                begin
                    inc(c[i], c[i-1]);
                    inc(d[x[i]], c[i] * w[i]);
                    dec(d[y[i]+1], c[i]*w[i]);
                end;
            for i:=1 to n do
                begin
                    inc(d[i], d[i-1]);
                    inc(a[i], d[i]);
                    S[i] := S[i-1] + a[i];
                end;
            read(u1, v1, S1);
            read(u2, v2, S2);
            if abs(S[v1] - S[u1-1] - S1) < abs(S[v2] - S[u2-1] - S2) then write('SAN WIN')
            else if abs(S[v1] - S[u1-1] - S1) > abs(S[v2] - S[u2-1] - S2) then write('MI WIN')
            else write('HOA');
end.
    
